motionplan/plan.go

```diff
@@ -0,0 +1,154 @@
+package motionplan
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	pb "go.viam.com/api/service/motion/v1"
+	"go.viam.com/rdk/motionplan/ik"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type Plan struct {
+	Trajectory
+	Path
+
+	// nodes corresponding to inputs can be cached with the Plan for easy conversion back into a form usable by RRT
+	// depending on how the Trajectory is constructed these may be nil and should be computed before usage
+	nodes []node
+}
+
+func newPlan(solution []node, sf *solverFrame, relative bool) (*Plan, error) {
+	traj := sf.nodesToTrajectory(solution)
+	path, err := newPath(solution, sf)
+	if err != nil {
+		return nil, err
+	}
+	return &Plan{
+		Trajectory: traj,
+		Path:       path,
+		nodes:      solution,
+	}, nil
+}
+
+// TODO: can probably think of a better name for this function
+func (plan *Plan) Remaining(waypointIndex int) *Plan {
+	// TODO: I don't think a deep copy should be necessary here but maybe?
+	return &Plan{
+		Path:       plan.Path[waypointIndex:],
+		Trajectory: plan.Trajectory[waypointIndex:],
+		nodes:      plan.nodes[waypointIndex:],
+	}
+}
+
+type Trajectory []map[string][]referenceframe.Input
+
+// GetFrameInputs is a helper function which will extract the waypoints of a single frame from the map output of a Trajectory.
+func (traj Trajectory) GetFrameInputs(frameName string) ([][]referenceframe.Input, error) {
+	solution := make([][]referenceframe.Input, 0, len(traj))
+	for _, step := range traj {
+		frameStep, ok := step[frameName]
+		if !ok {
+			return nil, fmt.Errorf("frame named %s not found in Trajectory", frameName)
+		}
+		solution = append(solution, frameStep)
+	}
+	return solution, nil
+}
+
+// String returns a human-readable version of the Trajectory, suitable for debugging.
+func (traj Trajectory) String() string {
+	var str string
+	for _, step := range traj {
+		str += "\n"
+		for component, input := range step {
+			if len(input) > 0 {
+				str += fmt.Sprintf("%s: %v\t", component, input)
+			}
+		}
+	}
+	return str
+}
+
+// Evaluate assigns a numeric score to a plan that corresponds to the cumulative distance between input waypoints in the Trajectory.
```

Comment:
[nit/opt]: `cumulative distance` what are the units?

Reply:
This is unitless.  It should change from being a "cumulative distance" to a more general score